encoding/migration/convert.go

```diff
@@ -0,0 +1,125 @@
+
+package migration
+
+import (
+	"fmt"
+
+	"github.com/m3db/m3metrics/encoding"
+	"github.com/m3db/m3metrics/metadata"
+	"github.com/m3db/m3metrics/metric/unaggregated"
+	"github.com/m3db/m3metrics/policy"
+)
+
+func toUnaggregatedMessageUnion(
+	metric unaggregated.MetricUnion,
+	policiesList policy.PoliciesList,
+) (encoding.UnaggregatedMessageUnion, error) {
+	metadatas := toStagedMetadatas(policiesList)
+	switch metric.Type {
+	case unaggregated.CounterType:
+		return encoding.UnaggregatedMessageUnion{
+			Type: encoding.CounterWithMetadatasType,
+			CounterWithMetadatas: unaggregated.CounterWithMetadatas{
+				Counter:         metric.Counter(),
+				StagedMetadatas: metadatas,
+			},
+		}, nil
+	case unaggregated.BatchTimerType:
+		return encoding.UnaggregatedMessageUnion{
+			Type: encoding.BatchTimerWithMetadatasType,
+			BatchTimerWithMetadatas: unaggregated.BatchTimerWithMetadatas{
+				BatchTimer:      metric.BatchTimer(),
+				StagedMetadatas: metadatas,
+			},
+		}, nil
+	case unaggregated.GaugeType:
+		return encoding.UnaggregatedMessageUnion{
+			Type: encoding.GaugeWithMetadatasType,
+			GaugeWithMetadatas: unaggregated.GaugeWithMetadatas{
+				Gauge:           metric.Gauge(),
+				StagedMetadatas: metadatas,
+			},
+		}, nil
+	default:
+		return encoding.UnaggregatedMessageUnion{}, fmt.Errorf("unknown metric type: %v", metric.Type)
+	}
+}
+
+// TODO: look into reuse metadatas during conversion.
+func toStagedMetadatas(
+	policiesList policy.PoliciesList,
+) metadata.StagedMetadatas {
+	numStagedPolicies := len(policiesList)
+	res := make(metadata.StagedMetadatas, 0, numStagedPolicies)
+	for _, sp := range policiesList {
+		if sp.IsDefault() {
+			sm := metadata.DefaultStagedMetadata
+			res = append(res, sm)
+			continue
+		}
+		sm := metadata.StagedMetadata{}
+		sm.CutoverNanos = sp.CutoverNanos
+		sm.Tombstoned = sp.Tombstoned
+		policies, _ := sp.Policies()
+		sm.Metadata = toMetadata(policies)
+		res = append(res, sm)
+	}
+	return res
+}
+
+// TODO: look into reuse metadata during conversion.
+func toMetadata(policies []policy.Policy) metadata.Metadata {
+	res := metadata.Metadata{}
+	for _, p := range policies {
+		// Find if there is an existing pipeline in the set of metadatas
+		// with the same aggregation ID.
+		pipelineIdx := -1
+		for i := 0; i < len(res.Pipelines); i++ {
+			if p.AggregationID == res.Pipelines[i].AggregationID {
+				pipelineIdx = i
+				break
+			}
+		}
+		if pipelineIdx == -1 {
+			res.Pipelines = append(res.Pipelines, metadata.PipelineMetadata{
+				AggregationID: p.AggregationID,
+			})
+			pipelineIdx = len(res.Pipelines) - 1
+		}
+
+		// Find if the storage policy already exists in the set of storage
+		// policies in the pipeline.
+		pipelines := res.Pipelines
+		policyIdx := -1
+		for i := 0; i < len(pipelines[pipelineIdx].StoragePolicies); i++ {
+			if pipelines[pipelineIdx].StoragePolicies[i] == p.StoragePolicy {
+				policyIdx = i
+				break
+			}
+		}
+		if policyIdx != -1 {
```

Comment:
When does this case happen? Is it only when there are duplicated policy in the `policies` list?

Reply:
Yes, granted it shouldn't happen so this is just defensive programming.